### ISSUE TITLE
Unexpected behavior of sensu-client in a redundant scenario

### ISSUE THREAD

**Issue by manupathak** [sensu/sensu#1044]
Hi,

I am seeing some possibly unexpected behavior with the sensu client which may or may not be a bug and wanted to see if somebody here may be able to answer my query. If this is not the right place, I apologize in advance; please let me know what would be the right forum for posting this query and I will post it there.

I am trying to design and validate a redundant Sensu solution that will across multiple data centers. I have already gone through the document on the Sensu website that talks about the different approaches:

https://sensuapp.org/docs/0.16/scaling_strategies

Given the project requirements and resources available, among other things, the approach I have settled on is to have two different sets of Sensu server/Rabbit/Redis instances, one in each data center (let's call them sensu-1, rabbit-1, redis-1 and sensu-2, rabbit-2, redis-2 respectively) and have the clients point to a sensu/rabbit/redis hostname that resolves to sensu-1, rabbit-1 and redis-1 when everything is working in data center #1, and to sensu-2, rabbit-2 and redis-2 if anything goes wrong in data center #1. This is essentially an ACTIVE-PASSIVE solution, with the Sensu setup in data center #2 taking over whenever there is a problem in data center #1.

In order to test this, I am modifying the DNS entries of sensu/rabbit/redis hostnames that the clients are using to point to the IP addresses of sensu-2, rabbit-2 and redis-2 (from those of sensu-1, rabbit-1 and redis-1, which is what they are pointing to initially). I would expect that once this change has been made, the clients would start using sensu-2, rabbit-2 and redis-2, and I should be able to see these clients as belonging to datacenter #2, instead of datacenter #1. 

However, in my testing I have noticed that unless I restart the sensu-client on all the nodes being monitored, this change does not take effect. In other words, I can verify on the client system that the sensu/rabbit/redis hostnames are now resolving to sensu-2, rabbit-2 and redis-2 after making the changes to DNS, but the sensu client is still 

- Is this behavior expected or is it a bug? Do I need any additional configuration to make this work?
- If this is expected behavior, how do I design a redundant solution that works across data centers that does not require restarting the clients in case of an outage?

Thanks in advance!

**Comment by portertech** [comment]
@manupathak sounds like the Sensu process is caching the DNS lookup result. We need to research manually flushing cached DNS lookup before reconnect.

**Comment by portertech** [comment]
This may require sensu-transport improvements, to handle or flush dns lookups.

**Comment by portertech** [comment]
https://github.com/sensu/sensu-transport

**Comment by manupathak** [comment]
Hi,

Thanks for getting back to me! Yes, I can confirm that this is due to DNS caching. For now, I was able to get around this issue by having a script check the IPs to which the sensu/rabbit/redis servers are resolving, and if they change, restarting the Sensu client. This works, but it is less than ideal, because it requires modifying all the Sensu clients, and still does not solve the problem of the Sensu clients showing up in the dashboard for the datacenter with the problem -- the client entries stay there until they are manually deleted from the dashboard.

What would be ideal is the following could be implemented:

- A config option which specifies if the DNS lookup results should be cached by the client. If this is set to no, the clients will always attempt to resolve the hostname before establishing a connection.

- A way for the clients to deregister and have Uchiwa take that into account, so that stale entries don't show up on the dashboard.

I would say either of these should not be hard to implement, and they would make Sensu a lot more attractive for anybody considering deploying it, since having to make modifications on every client may not be always be an option.

I don't know what your priorities are, but I would be interested in hearing your thoughts on this.

Thanks!
-Manu

**Comment by calebhailey** [comment]
Interesting issue. I wondered if perhaps Sensu's ability to accept an array of RabbitMQ endpoints would satisfy this issue, but according to the [Sensu documentation on RabbitMQ configuration][1], Sensu will randomly sort those RabbitMQ endpoints, so you wouldn't have control over the priority. 

I wonder if we could add support for a transport "fallback" with a configurable threshold for when the Sensu client should give up on trying to reconnect to the primary (active) transport, and fallback to connecting to the `fallback` (passive) transport. 

Some questions this raises for me are: 

- Would it make sense to implement said `fallback` support for the transport only, and not also for the [data store][2] and API (i.e. only useful for the Sensu client)? Or would an active/passive `fallback` configuration need to provide passive configuration for the entire Sensu stack? 
- Could the current behavior be related to underlying Ruby DNS behavior? Does it make more sense to think of this as a feature request for "active/passive transport support for Sensu" rather than a bug? 

[1]: https://sensuapp.org/docs/latest/rabbitmq#configuring-sensu-for-a-rabbitmq-cluster
[2]: https://sensuapp.org/docs/latest/data-store

**Comment by manupathak** [comment]
Hi,

Thanks for getting back on this. As I mentioned earlier, I ended up implementing a script that runs on each client host, and that restarts the Sensu client whenever it detects a change in the DNS entry, and this addresses the issue I had raised. However, the fact that this has to be run on every single client is not so great.

1. The fallback option you are suggesting would work, but I don't know how many deployments of Sensu would want a fallback option to support the solution I describe. If the Sensu client were instead modified to just flush the DNS cache with a new option, I think it might be more useful. Of course, I would be happy to see either of those integrated into the current codebase.

2. I think the current behavior is really part of the transport component of Sensu, as Sean pointed out. Whatever the change might be that's implemented to support the behavior I would like to see is probably a small enhancement, so this is probably not a bug. However, I do  think it is a shortcoming of the Sensu client that supporting redundancy in this scenario requires running something on each host to restart the client explicitly, rather than have that happen automatically.

-Manu
